src/predictor/array_tree_layout.h

```diff
@@ -0,0 +1,300 @@
+/**
+ * Copyright 2021-2025, XGBoost Contributors
+ * \file array_tree_layout.cc
+ * \brief Implementation of array tree layout -- a powerfull inference optimization method.
+ */
+#ifndef XGBOOST_PREDICTOR_ARRAY_TREE_LAYOUT_H_
+#define XGBOOST_PREDICTOR_ARRAY_TREE_LAYOUT_H_
+
+#include <limits>
+#include <vector>
+
+#include "predict_fn.h"             // for GetNextNode, GetNextNodeMulti
+#include "xgboost/tree_model.h"     // for RegTree, MTNotImplemented, RTreeNodeStat
+
+namespace xgboost::predictor {
+
+/**
+ * @brief The class holds the array-based representation of the top levels of a single tree.
+ * 
+ * @tparam TreeType The type of the origianl tree (RegTree or MultiTargetTree)
+ *
+ * @tparam has_categorical if the tree has categorical features
+ *
+ * @tparam any_missing if the class is able to process missing values
+ * 
+ * @tparam kNumDeepLevels number of tree leveles being unrolled into array-based structure
+ */
+template <class TreeType, bool has_categorical, bool any_missing, int kNumDeepLevels>
+class ArrayTreeLayout {
+ private:
+  /* Number of nodes in the array based representation of the top levels of the tree
+   */
+  constexpr static size_t kNodesCount = (1u << kNumDeepLevels) - 1;
+
+  struct Empty {};
+  using DefaultLeftType =
+        typename std::conditional_t<any_missing,
+                                   std::array<uint8_t, kNodesCount>,
+                                   struct Empty>;
+  using IsCatType =
+        typename std::conditional_t<has_categorical,
+                                   std::array<uint8_t, kNodesCount>,
+                                   struct Empty>;
+  using CatSegmentType =
+        typename std::conditional_t<has_categorical,
+                                   std::array<common::Span<uint32_t const>, kNodesCount>,
+                                   struct Empty>;
+
+  DefaultLeftType default_left_;
+  IsCatType is_cat_;
+  CatSegmentType cat_segment_;
+
+  std::array<bst_feature_t, kNodesCount> split_index_;
+  std::array<float, kNodesCount> split_cond_;
+  /* The nodes at tree levels 0, 1, ..., kNumDeepLevels - 1 are unrolled into an array-based structure.
+   *  If the tree has additional levels, this array stores the node indices of the sub-trees at level kNumDeepLevels.
+   *  This is necessary to continue processing nodes that are not eligible for array-based unrolling.
+   *  The number of sub-trees packed into this array is equal to the number of nodes at tree level kNumDeepLevels,
+   *  which is calculated as (1u << kNumDeepLevels) == kNodesCount + 1.
+   */
+  std::array<bst_node_t, kNodesCount + 1> nidx_in_tree_;
+
+  static bool IsLeaf(const RegTree& tree, bst_node_t nidx) {
```

Comment:
Is there a benefit of doing this C++ overloading rather than the simpler `tree.IsLeaf`? How much faster are we seeing?